feat: GutenbergKit synchronizes featured images

dcalhoun opened the pull request:
## Description
<!-- Describe the changes, why they are needed, and how they address the issue -->
Handle featured image updates by updating associated post data, so that
the correct featured image is persisted and displayed in the post
settings UI.

Ref CMM-383. 

## Testing instructions
<!-- Consider listing specific testing steps for helping reviewers evaluate the efficacy of the changes -->
See https://github.com/wordpress-mobile/GutenbergKit/pull/147.

<!--

Consider testing the following *yourself* before requesting a review:

- Compatibility with WordPress.com sites and self-hosted Jetpack sites
- Both portrait and landscape orientations
- Light and dark modes
- Dynamic font sizes: larger, smaller and bold text
- High contrast mode
- Screen reader experience (e.g., VoiceOver)
- Languages with large words or with letters/accents not frequently used in English
- Right-to-left languages layout
- Multi-tasking: split view and slide over

-->

dcalhoun committed:
feat: GutenbergKit synchronizes featured images

Handle featured image updates by updating associated post data, so that
the correct featured image is persisted and displayed in the post
settings UI.

dangermattic commented:
<!--
  0 Errors
  0 Warnings
  1 Message: This PR is still a Draft: some...
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">This PR is still a Draft: some checks will be skipped.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

kean commented on WordPress/Classes/ViewRelated/NewGutenberg/NewGutenbergFeaturedImageHelper.swift:
```diff
@@ -0,0 +1,22 @@
+import Foundation
+import WordPressShared
+
+class NewGutenbergFeaturedImageHelper: NSObject {
```
(nit) Convert to a simple `struct`.

kean commented on WordPress/Classes/ViewRelated/NewGutenberg/NewGutenbergViewController.swift:
```diff
@@ -27,6 +27,10 @@ class NewGutenbergViewController: UIViewController, PostEditor, PublishingEditor
         return GutenbergMediaPickerHelper(context: self, post: post)
     }()
 
+    lazy var featuredImageHelper: NewGutenbergFeaturedImageHelper = {
```
(nit) `lazy let featuredImageHelper = NewGutenbergFeaturedImageHelper(post: post)`

kean commented on WordPress/Classes/ViewRelated/NewGutenberg/NewGutenbergViewController.swift:
```diff
@@ -417,6 +421,17 @@ extension NewGutenbergViewController: GutenbergKit.EditorViewControllerDelegate
+    func editor(_ viewController: GutenbergKit.EditorViewController, didUpdateFeaturedImage mediaID: Int) {
+        let featuredImageID = post.featuredImage?.mediaID
+
+        guard featuredImageID as? Int != mediaID else {
```
(nit) I'd suggest using `post.featuredImage?.mediaID?.intValue` instead. I'm not sure `featuredImageID as? Int` is the correct conversion from `NSNumber` to `Int`.

kean reviewed (approved):
Worked as expected.
I haven't tested the scenario where the image doesn't change.

kean commented on WordPress/Classes/ViewRelated/NewGutenberg/NewGutenbergFeaturedImageHelper.swift:
```diff
@@ -0,0 +1,22 @@
+    }
+
+    func setFeaturedImage(mediaID: Int) {
+        let media = Media.existingMediaWith(mediaID: NSNumber(value: mediaID), inBlog: post.blog)
```
Sorry, missed this – this part fails when uploading new media via the new editor. It only works for the media that's already synced on the device. I believe what you need to do is use `existingOrStubMediaWith`. It also needs to save the Core Data context. I can help look into it tmrw.

kean replied:
Example usage in ObjC: https://github.com/wordpress-mobile/WordPress-iOS/blob/98256ec18da9227b9bdb228cb8c94f7067bdf1fe/Sources/WordPressData/Objective-C/PostHelper.m#L37.

dcalhoun replied:
I do not follow your concern. From my testing, the current implementation works when setting an image as the featured image when:

1. Inserting an image from the Media Library;
2. And uploading a new image in the editor. 

When using "Insert from URL" or if an upload is in-progress, the "Set as featured image" button is not displayed. 

What am I overlooking? Is a change still needed?

kean replied:
The steps to reproduce the issue:

https://github.com/user-attachments/assets/81ac2a27-9d39-49ad-b0da-690edb870dbf

Observed Result: "Post Settings" fails to show the featured image when it's uploaded by the editor (web view).

I haven't stepped through the code with a debugger, but presumably, it happens because when this line is called, the app doesn't have an existing `Media` entity in the database:

```swift
func setFeaturedImage(mediaID: Int) {
         let media = Media.existingMediaWith(mediaID: NSNumber(value: mediaID), inBlog: post.blog)
```

The app needs to use `existingOrStubMediaWith` instead. If the app hasn't synced the `Media` entity yet, it will create a placeholder and will populate it later when it needs to display it.

dcalhoun replied:
Interesting. I haven't reproduced yet. Thanks for sharing, I'll debug some more.

dcalhoun replied:
I was able to reproduce it this time. I'm guessing I previously confused myself while toggling your suggested fix on/off. 

Resolved in https://github.com/wordpress-mobile/WordPress-iOS/pull/24588/commits/c2f51a55e00b1a6d791879d3a95b50cbfa6af4ec. Thanks again for the suggestion.

dcalhoun committed:
refactor: Simplify class to a struct

dcalhoun committed:
fix: Ensure newly uploaded featured images display

The stub serves as a placeholder and is populated later once the app
needs to display the image.

dcalhoun committed:
build: Update GutenbergKit version
